trade.lic

```diff
@@ -1558,126 +1646,233 @@ class Trade
+    return unless @current_materials['balsa lumber'] < @lumber_quantity
+    return unless time_to_room(Room.current.id, @crafting_data['shaping'][town]['stock-room']) < 20.0
     be_outside_caravan
     crafting_data = get_data('crafting')
     stow_hands
+    shift_hometown(town)
     ensure_copper_on_hand(3000, @settings) unless skipcoin
-    walk_to(crafting_data['shaping'][closest_town]['stock-room'])
+    walk_to(crafting_data['shaping'][town]['stock-room'])
     bput('get my balsa lumber', 'You get', 'You are already', 'What were you')
-    order_item(crafting_data['shaping'][closest_town]['stock-room'], 11)
+    order_item(crafting_data['shaping'][town]['stock-room'], 11)
     bput('combine', 'combine')
-    @current_lumber_count = bput('count my balsa lumber', 'You count out \d+').scan(/\d+/).first.to_i
-    buy_wood(true) if @current_lumber_count < @lumber_quantity
+    count_materials('balsa lumber')
+    buy_wood(true) if @current_materials['balsa lumber'] < @lumber_quantity
     stow_hands
+    clear_hometown
   end
 
-  def buy_yarn(skipcoin = false)
+  def buy_outfitting_mats
     return unless @caravan_training_skills.include?('Outfitting')
-    return unless @crafting_data['tailoring'][closest_town]
-    return unless @current_yarn_count < @yarn_quantity
-    return unless time_to_room(Room.current.id, @crafting_data['shaping'][closest_town]['stock-room']) < 20.0
+    town = closest_town
+    return unless @crafting_data['tailoring'][town]
+    return unless time_to_room(Room.current.id, @crafting_data['shaping'][town]['stock-room']) < 20.0
     be_outside_caravan
-    crafting_data = get_data('crafting')
+    shift_hometown(town)
+    crafting_data = get_data('crafting')['tailoring'][town]
     stow_hands
-    ensure_copper_on_hand(3000, @settings) unless skipcoin
-    walk_to(crafting_data['tailoring'][closest_town]['stock-room'])
+    ensure_copper_on_hand(5000, @settings)
+
+    should_knit? ? buy_yarn(crafting_data) : buy_cloth(crafting_data)
+    clear_hometown
+  end
+
+  def should_knit?
+    @crafting_override['tailoring']['recipe'].include?('knitted') || DRSkill.getrank('Outfitting') < 550
+  end
+
+  def buy_yarn(crafting_data)
+    return unless @current_materials[@outfitting_material_type] < @outfitting_quantity
+    walk_to(crafting_data['stock-room'])
     bput('get my wool yarn', 'You get', 'You are already', 'What were you')
-    order_item(crafting_data['tailoring'][closest_town]['stock-room'], 13)
+    order_item(crafting_data['stock-room'], 13)
     bput('combine my yarn', 'You combine', 'You must be holding both')
-    @current_yarn_count = bput('count my wool yarn', 'You count out \d+ yards').scan(/\d+/).first.to_i
-    buy_yarn(true) if @current_yarn_count < @yarn_quantity
+    count_materials('wool yarn')
+    buy_yarn(crafting_data) if @current_materials[@outfitting_material_type] < @outfitting_quantity
+    stow_hands
+  end
+
+  def buy_cloth(crafting_data)
```

Comment:
❓ Should any of this logic be in common-crafting?

Reply:
Yeah, the plan is definitely to commonize some of this and make it work for ;craft.  But I don't yet know the scope of that.  I'll look at that later.